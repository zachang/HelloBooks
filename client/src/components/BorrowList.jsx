import React from 'react';
import PropTypes from 'react-proptypes';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import swal from 'sweetalert2';
import { viewAllBorrowAction, confirmBorrowAction }from '../actions/bookAction';
import AdminHeader from './common/AdminHeader';
import AdminSidebar from './common/AdminSidebar';
import { Pagination } from 'react-materialize';
import UserBorrowList from './borrow/UserBorrowList.jsx';

/**
 * BorrowList class declaration
 * @class BorrowList
 * @extends {React.Component}
 */
export class BorrowList extends React.Component {
  /**
   * class constructor
   * @param {object} props
   */
  constructor(props){
    super(props);
    this.state = {
      errors: null,
      pageCount: null,
      limit: 15,
      clickedBorrowList: []
    };
    this.confirmBorrow = this.confirmBorrow.bind(this);
  }

  /**
   * @method componentWillMount
   * @return {void} void
   */
  componentWillMount() {
    this.props.viewAllBorrowAction(this.state.limit, 0);
  }

  /**
   * @method componentWillReceiveProps
   * @param {object} nextProps - nextProps
   * @return {object} nextProps
   */
  componentWillReceiveProps(nextProps) {
    this.setState({ clickedBorrowList: nextProps.bookState.clickedBorrowList });
    if(nextProps.bookState.success === false){
      this.setState({ errors: nextProps.bookState.errors });
    }
    if (this.state.pageCount !== nextProps.bookState.pageCount) {
      this.setState({ pageCount: nextProps.bookState.pageCount });
    }
  }

  /**
   * Handles  book borrow confirmation
   * @method confirmBorrow
   * @return {void}
   * @param {integer} borrowId - borrowId
   */
  confirmBorrow(borrowId) {
    swal({
      title: 'Are you sure?',
      text: "You won't be able to revert this!",
      type: 'warning',
      successButtonColor: '#26a69a',
      showCancelButton: true,
      confirmButtonColor: '#26a69a',
      cancelButtonColor: '#d33',
      confirmButtonText: 'Confirm!'
    }).then((result) => {
      if (result.value) {
        swal(
          'Confirmed!',
          'Borrow has been confirmed',
          'success'
        );
        this.props.confirmBorrowAction(borrowId);
      }
    });
    this.setState({ showToast: true });
  }

  /**
   * Renders BorrowList component
   * @return {XML} JSX
   */
  render() {
    return (
      <div className='row'>
        <AdminHeader/>
        <AdminSidebar/>
        <div className='container mainCon' style={{marginLeft: '5%'}}>
          <div className='row'>
            <div className='section'>
              <h4 style={{ marginTop: '7%' }}>Borrowed Books</h4>
            </div>
            <div className='divider' style={{width: '100%', marginTop: '-2.5%', marginBottom: '3.5%'}}></div>

            <div className='row'>
              <table className='bordered highlight'>
                <thead className='black white-text'>
                <tr>
                  <th>Book Name</th>
                  <th>Category</th>
                  <th>Borrower</th>
                  <th>Date</th>
                  <th>Return_status</th>
                  <th>Confirm_Borrow</th>
                </tr>
                </thead>

                <tbody>
                { this.props.bookState.borrowers.map((borrower, i) =>
                    <UserBorrowList
                      key={i}
                      borrower={ borrower }
                      clickedBorrowList={this.state.clickedBorrowList}
                      confirmBorrow={this.confirmBorrow}
                    />
                )}
                </tbody>
              </table>
            </div>
            <div className='row'>
              {
                ((this.state.pageCount) ?
                  <Pagination
                    items={this.state.pageCount}
                    onSelect={(page) => {
                      const offset = (page - 1) * this.state.limit;
                      this.props.viewAllBorrowAction(this.state.limit, offset);
                    }} /> : '')
              }
            </div>

          </div>
        </div>
      </div>

    );
  }
}


BorrowList.propTypes = {
  bookState: PropTypes.object.isRequired,
  viewAllBorrowAction: PropTypes.func.isRequired,
  confirmBorrowAction: PropTypes.func.isRequired
};
const mapStateToProps = state => ({
  bookState: state.bookReducer,
});
const mapDispatchToProps = dispatch =>
  bindActionCreators({ viewAllBorrowAction, confirmBorrowAction }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(BorrowList);