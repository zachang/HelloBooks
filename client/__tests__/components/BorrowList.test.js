import React from 'react';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import storeConfiguration from '../../src/store/store';
import { BorrowList } from '../../src/components/BorrowList.jsx';

describe('<BorrowList/>', () => {
  // create a spy function for viewAllBorrowAction
  const viewAllBorrowAction = sinon.spy();
  // create a spy function for confirmBorrowAction
  const confirmBorrowAction = sinon.spy();
  // spy on confirmReturn of BorrowList Page
  sinon.spy(BorrowList.prototype, 'confirmBorrow');
  // spy on componentWillMount of BorrowList Page
  sinon.spy(BorrowList.prototype, 'componentWillMount');
  // spy on componentWillReceiveProps of BorrowList Page
  sinon.spy(BorrowList.prototype, 'componentWillReceiveProps');

  const props = {
    bookState: {
      success: false,
      passes: null,
      errors: null,
      fails: null,
      books: [],
      book: null,
      borrows: null,
      allBorrows: null,
      borrowers: ['mmmm'],
      returned: null,
      returnings: null,
      returners: ['mmmm'],
      pageCount: null,
      confirmReturn: null,
      clickedReturnedBorrowList: [],
      confirmBorrow: null,
      clickedBorrowList: []
    },
    showToast: true,
    viewAllBorrowAction,
    confirmBorrowAction
  };

  const middleware = [thunk];
  const mockStore = configureMockStore(middleware);
  let store;
  let shallowComponent;
  beforeEach(() => {
    store = mockStore(storeConfiguration);
    shallowComponent = shallow(<BorrowList {...props} />);
    shallowComponent.setState({
      pageCount: 1,
      borrowers: ['mmm'],
      showToast: true,
      clickedBorrowList: []
    });
  });

  it('renders <BorrowList /> component', () => {
    const tree = toJson(shallowComponent);
    expect(tree).toMatchSnapshot();
  });

  it('calls componentWillReceiveProps when paginating', () => {
    props.pageCount = 1;
    props.bookState.pageCount = 2;
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(BorrowList.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it(`should check if the 'Pagination' is defined`, () => {
    expect(shallowComponent.find('Pagination').length).toBe(1);
    shallowComponent.find('Pagination').simulate('select');
  });

  it('calls componentWillReceiveProps when paginating', () => {
    props.bookState.sucess = false;
    shallowComponent.setState({
      errors: props.bookState.errors
    });
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(BorrowList.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it('should check that confirmReturn is called', () => {
    shallowComponent.instance().confirmBorrow(1);
    expect(BorrowList.prototype.confirmBorrow.called).toEqual(true);
  });
});