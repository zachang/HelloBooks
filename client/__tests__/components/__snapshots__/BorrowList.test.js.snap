// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BorrowList/> renders <BorrowList /> component 1`] = `
<div
  className="row"
>
  <div
    className="section"
  >
    <h4
      style={
        Object {
          "marginTop": "7%",
        }
      }
    >
      Borrowed Books
    </h4>
  </div>
  <div
    className="divider"
    style={
      Object {
        "marginBottom": "3.5%",
        "marginTop": "-2.5%",
        "width": "100%",
      }
    }
  />
  <div
    className="row"
  >
    <table
      className="bordered highlight"
    >
      <thead
        className="black white-text"
      >
        <tr>
          <th>
            Book Name
          </th>
          <th>
            Category
          </th>
          <th>
            Borrower
          </th>
          <th>
            Date
          </th>
          <th>
            Return_status
          </th>
          <th>
            Confirm_Borrow
          </th>
        </tr>
      </thead>
      <tbody>
        <userBorrowList
          borrower="mmmm"
          clickedBorrowList={Array []}
          confirmBorrow={[Function]}
          key="0"
        />
      </tbody>
    </table>
  </div>
  <div
    className="row"
  >
    <Pagination
      activePage={1}
      items={1}
      onSelect={[Function]}
    />
  </div>
</div>
`;
